lib/elixir/src/elixir_expand.erl

```diff
@@ -406,6 +406,7 @@ expand({Name, Meta, Kind} = Var, E) when is_atom(Name), is_atom(Kind) ->
 
 expand({Atom, Meta, Args}, E) when is_atom(Atom), is_list(Meta), is_list(Args) ->
   assert_no_ambiguous_op(Atom, Meta, Args, E),
+  assert_no_clauses(Atom, Meta, Args, E),
```

Comment:
Aren't we going to have false positives by moving it here? Imagine we have this macro:

```elixir
def escape(quoted) do
  Macro.escape(quoted)
end
```

This should still work:

```elixir
escape do
  :bar
end
```

So the problem here is really when we are sure we have a local (as in expand_local) or a remote (expand_remote) and it contains a `:->` element regardless of the key! By the simple fact it isn't macro, it means `->` won't compile.